### PR TITLE
Implement config parsing with Pydantic and OmegaConf

### PR DESCRIPTION
**Pull request by pedromsferraz-mila**
This PR introduces the configuration and CLI infrastructure needed to run full active learning experiments from YAML files.

#### Pydantic Configuration Layer

Each framework component now has a corresponding Pydantic config class in `<component>/config.py`. Each config class follows the same pattern:

- A `type` field (using `Literal`) acts as a discriminator so Pydantic can resolve the correct subclass from the YAML `type` key.
- A `build()` method instantiates the corresponding runtime object from the validated config.
- A top-level `ActiveLearningConfig` model in `src/activelearning/config.py` composes all per-component configs into a single validated schema.

#### OmegaConf Loading Pipeline

The `utils/config_loader.py` module manages the load-merge-validate pipeline:
1. OmegaConf loads and merges one or more YAML files from left to right. Later files overwrite earlier files for shared keys.
2. The loader merges CLI dotlist overrides (e.g., `budget.available_budget=10`) on top of the YAML configurations.
3. Pydantic validates the final merged dictionary against `ActiveLearningConfig` using `model_validate`.

This design supports modular configurations, enabling researchers to separate shared runtime blocks from experiment-specific overrides.

#### Entry point

The 'main.py' script connects the CLI to the configuration pipeline. Executing the command:
```bash
uv run activelearning <config.yaml> [<override.yaml> ...] [key=value ...]
``` 
loads the configurations, instantiates all components via their `build()` methods, and initializes the active learning loop.

#### Example configuration

`config/branin_toy_example.yaml` provides a complete, runnable baseline example: BoTorch GP surrogate, multi-fidelity acquisition with a hypercube candidate set, HypercubeSampler with three fidelity levels, CostAwareSelector, BraninOracle with fidelity costs, a constant round budget schedule, and ConsoleLogger.

### CODE REVIEW

File: src/activelearning/main.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+def _parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(
+        prog="activelearning",
+        description="Run the active learning loop from a YAML config file.",
+    )
+    parser.add_argument(
+        "config",
+        nargs="+",
+        help="Path(s) to YAML config file(s). Multiple files are merged left to right.",
+    )
```
[the remark this answers is not in the document]
Comment (pedromsferraz-mila) [reply src/activelearning/main.py:54]: Good catch! Fixed: [78f1b37](https://github.com/milaforscience/activelearning/pull/8/commits/78f1b376fa67efe4170f4acdc48c617d7cf37a0c)

File: src/activelearning/utils/settings.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    output_dir: Path = PROJECT_ROOT / "output"
+
+    # Load environment variables from .env files
+    model_config: ClassVar = SettingsConfigDict(
+        env_file=ENV_FILES, env_ignore_empty=True, extra="ignore"
+    )
+
+
+# Instantiate a global singleton
+settings = GlobalSettings()
```
[the remark this answers is not in the document]
Comment (pedromsferraz-mila) [reply src/activelearning/utils/settings.py]: Good point, this file is no longer useful. Removed in [b731036](https://github.com/milaforscience/activelearning/pull/8/commits/b73103643a6c7c8ca87b86eb3d89c8d03af8f3c0)

File: src/activelearning/utils/settings.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+from pathlib import Path
+from typing import ClassVar
+from pydantic_settings import BaseSettings, SettingsConfigDict
+
+# Define paths to environment files
+CURRENT_FILE = Path(__file__).resolve()
+PACKAGE_ROOT = CURRENT_FILE.parent.parent.parent.parent
+PROJECT_ROOT = PACKAGE_ROOT
```
[the remark this answers is not in the document]
Comment (pedromsferraz-mila) [reply src/activelearning/utils/settings.py]: Removed `settings.py` in [b731036](https://github.com/milaforscience/activelearning/pull/8/commits/b73103643a6c7c8ca87b86eb3d89c8d03af8f3c0).

File: src/activelearning/budget/schedule_config.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+def sigmoid_iteration_schedule(
+    total_budget: float,
+    num_iterations: int,
+    midpoint_fraction: float,
+    steepness: float,
+):
+    def _sigmoid(x: float) -> float:
+        return 1.0 / (1.0 + math.exp(-steepness * (x - midpoint_fraction)))
+
```
[the remark this answers is not in the document]
Comment (pedromsferraz-mila) [reply src/activelearning/budget/schedule_config.py]: Good point, fixed using torch's sigmoid: 61963c70cbd957f017863d8f2202207f79fbb50a

File: src/activelearning/budget/schedule_config.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+from typing import Annotated, Literal, Union
+
+from pydantic import BaseModel, Field
+
+
+def constant_schedule(value: float):
+    def schedule(_: int) -> float:
+        return value
+
+    return schedule
```
[the remark this answers is not in the document]
Comment (pedromsferraz-mila) [reply src/activelearning/budget/schedule_config.py]: Good point, fixed: [a540306](https://github.com/milaforscience/activelearning/pull/8/commits/a5403067a95f693730ea78a7fd684dce7a31df9f)

File: tests/utils/test_config_loader.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+def test_load_config_single_file(base_config, acquisition_config):
+    """Single-file load should not raise and must contain the base keys."""
+    # Merge both files so the config is complete (acquisition required)
+    cfg = load_config([base_config, acquisition_config])
+    assert "dataset" in cfg
+    assert "surrogate" in cfg
+    assert "budget" in cfg
+    assert "acquisition" in cfg
+
```
[the remark this answers is not in the document]
Comment (pedromsferraz-mila) [reply tests/utils/test_config_loader.py]: Good point, fixed: d2ba921b8239595159f958b4f630a7c98910d12b

File: src/activelearning/budget/schedule_config.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    def _sigmoid(x: float) -> float:
+        return torch.sigmoid(torch.tensor(steepness * (x - midpoint_fraction))).item()
+
+    # Use CDF increments so cumulative spend follows a sigmoid.
+    weights = [
+        _sigmoid((i + 1) / num_iterations) - _sigmoid(i / num_iterations)
+        for i in range(num_iterations)
+    ]
+    weight_sum = sum(weights)
+    allocations = [total_budget * weight / weight_sum for weight in weights]
```
[the remark this answers is not in the document]
Comment (pedromsferraz-mila) [reply src/activelearning/budget/schedule_config.py]: Good catch! Fixed: 557a9887706a95b0794304a183db2cc890b32901

File: tests/utils/test_config_loader.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+        "argv",
+        ["activelearning", str(base_config), str(acquisition_config), override],
+    ):
+        # If the override were silently dropped, the loop would run with the
+        # base budget (0.05) instead of the tighter one (0.001). Both are valid
+        # runs so we can't distinguish them here; the key assertion is that the
+        # override token is not treated as a file path (which would raise).
+        main()
+
+
```
[the remark this answers is not in the document]
Comment (pedromsferraz-mila) [reply tests/utils/test_config_loader.py]: Good point, here's an improved test function: [1d9444c](https://github.com/milaforscience/activelearning/pull/8/commits/1d9444c3221aff3f84c33c4aa03f699b74bb5705)

File: src/activelearning/main.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        overrides=overrides or None,
+    )
+    bootstrap_logger_backend_imports(OmegaConf.to_container(raw_cfg, resolve=False))
+
+    from activelearning.active_learning import active_learning
+    from activelearning.config import ActiveLearningConfig
+    from activelearning.utils.config_loader import parse_config
+    from activelearning.runtime import bind_runtime_context
+
+    cfg = parse_config(raw_cfg, ActiveLearningConfig)
```
Comment (alexhernandezgarcia) [inline src/activelearning/main.py:90]: Would it make sense to pack all these lines into a method, something like `process_arguments()`?

Comment (pedromsferraz-mila) [reply src/activelearning/main.py:90]: Good point, done! 230746d8bbbf7c49004f82f3245b496c34a90f8a

Comment (alexhernandezgarcia) [reply src/activelearning/main.py:90]: Thanks!

File: src/activelearning/main.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        import warnings
+
+        warnings.warn(
+            f"Unrecognised arguments ignored: {unknown}. "
+            "Pass OmegaConf overrides as positional key=value tokens, "
+            "e.g. budget.available_budget=10.",
+            stacklevel=2,
+        )
+
+    if not configs:
```
Comment (alexhernandezgarcia) [inline src/activelearning/main.py:75]: I suppose you've thought of this already, so I'd like to know your thoughts: why requiring at least one config instead of setting `config/base.yaml` as default, for example?

File: src/activelearning/budget/budget_schedule.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
+import torch
+from typing import Annotated, Callable, Literal, Union
+
+from pydantic import BaseModel, Field
+
+
+def constant_schedule(value: float) -> Callable[[int], float]:
```
Comment (alexhernandezgarcia) [inline src/activelearning/budget/budget_schedule.py:7]: Why are these schedule functions defined here in the config script? Is this the standard way with pydantic?

File: src/activelearning/utils/config_loader.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
+from pathlib import Path
+from typing import TypeVar, cast
+
+from omegaconf import DictConfig, OmegaConf
+from pydantic import BaseModel
+
+ModelT = TypeVar("ModelT", bound=BaseModel)
```
Comment (alexhernandezgarcia) [inline src/activelearning/utils/config_loader.py:7]: This is a bit mysterious to me.
- What is `ModelT`? 
- Would you mind explaining what `TypeVar("ModelT", bound=BaseModel)` does?

File: src/activelearning/utils/config_loader.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    if overrides:
+        cfg = OmegaConf.merge(cfg, OmegaConf.from_dotlist(overrides))
+    return cast(DictConfig, cfg)
+
+
+def parse_config(cfg: DictConfig, model: type[ModelT]) -> ModelT:
+    return model.model_validate(OmegaConf.to_container(cfg, resolve=True))
+
+
+def load_and_parse(
```
Comment (alexhernandezgarcia) [inline src/activelearning/utils/config_loader.py:43]: This isn't used yet, right?

File: src/activelearning/runtime.py [hunk 14]
```diff
@@ -18,6 +18,30 @@ class RuntimeContext:
...
+def resolve_torch_dtype(precision: int) -> torch.dtype:
+    """Map a floating-point precision setting to a torch dtype."""
+    if precision == 32:
+        return torch.float32
+    if precision == 64:
+        return torch.float64
+    raise ValueError(f"Unsupported precision {precision}. Expected 32 or 64.")
+
+
+class RuntimeConfig(BaseModel):
```
Comment (alexhernandezgarcia) [inline src/activelearning/runtime.py]: I'd suggest to rename this `RuntimeContextConfig` for consistency with how other `*Config` classes are named.

File: src/activelearning/runtime.py [hunk 15]
```diff
@@ -18,6 +18,30 @@ class RuntimeContext:
...
+    """Configuration for global torch runtime defaults."""
+
+    device: str = "cpu"
+    precision: Literal[32, 64] = 64
+
+    def build_context(self, logger: Logger | None = None) -> RuntimeContext:
```
Comment (alexhernandezgarcia) [inline src/activelearning/runtime.py]: Why is this method named `build_context()` instead of just `build()` as in other classes?

File: src/activelearning/main.py [hunk 16]
```diff
@@ -0,0 +1,134 @@
+"""Entry point for the active learning loop.
+
+Usage
+-----
+    python -m activelearning <config.yaml> [<config2.yaml> ...] [key=value ...]
```
Comment (alexhernandezgarcia) [inline src/activelearning/main.py]: Is this correct? Isn't it `uv run activelearning ...`?

File: src/activelearning/main.py (hunk 10 above)
Comment (alexhernandezgarcia) [reply src/activelearning/main.py:75]: I take this back. It is good to require a concrete config to specify a task rather than have an arbitrary default one (e.g. Branin).

File: README.md [hunk 17]
```diff
@@ -58,6 +58,82 @@ make test
...
+uv run activelearning config/branin_toy_example.yaml
+```
+
+### What the Example Does
+
+`config/branin_toy_example.yaml` runs a multi-fidelity active learning loop on the 2D Augmented Branin benchmark. Key components:
+
+| Component | Details |
+|-----------|---------|
+| **Oracle** | Negated Augmented Branin with 3 fidelity levels (costs: 0.01 / 0.1 / 1.0) |
```
Comment (alexhernandezgarcia) [inline README.md]:
```suggestion
| **Oracles** | Negated Augmented Branin with 3 fidelity levels (costs: 0.01 / 0.1 / 1.0) |
```

File: README.md [hunk 18]
```diff
@@ -58,6 +58,82 @@ make test
...
+
+Run the Branin toy example:
+
+```sh
+uv run activelearning config/branin_toy_example.yaml
+```
+
+### What the Example Does
+
+`config/branin_toy_example.yaml` runs a multi-fidelity active learning loop on the 2D Augmented Branin benchmark. Key components:
```
Comment (alexhernandezgarcia) [inline README.md]:
```suggestion
`config/branin_toy_example.yaml` runs a multi-fidelity active learning loop on the discretized (100x100) Augmented Branin benchmark. Key components:
```

File: README.md [hunk 19]
```diff
@@ -58,6 +58,82 @@ make test
...
+
+| Component | Details |
+|-----------|---------|
+| **Oracle** | Negated Augmented Branin with 3 fidelity levels (costs: 0.01 / 0.1 / 1.0) |
+| **Surrogate** | BoTorch GP (`SingleTaskMultiFidelityGP`) |
+| **Acquisition** | qMFLBMES (lower-bound multi-fidelity max-value entropy search) |
+| **Sampler** | Latin Hypercube Sampling over the 2D domain (10 000 candidates) |
```
Comment (alexhernandezgarcia) [inline README.md]:
```suggestion
| **Sampler** | Latin Hypercube Sampling over the 100x100 design space |
```

File: config/branin_toy_example.yaml [hunk 20]
```diff
@@ -0,0 +1,53 @@
+# Toy experiment: 2D Augmented Branin benchmark with the BoTorch GP surrogate
```
Comment (alexhernandezgarcia) [inline config/branin_toy_example.yaml]:
```suggestion
# Toy experiment: Discrete 100x100 Augmented Branin benchmark with the BoTorch GP surrogate
```

**Review by alexhernandezgarcia — DISMISSED** [review]
Sorry for the huge delay in reviewing this.

I have tried to understand all the pydantic stuff and I'm still far from understanding it well, so my review can only be quite shallow.

I have pushed one commit (2edd072c86790bc1317c7825394303a48d76dee6) with just comments that may be helpful for future developers without pydantic knowledge (like me). Feel free to revert or change.

I have dropped other minor suggestions and questions here and there. Otherwise, it all looks good!

File: src/activelearning/budget/budget_schedule.py (hunk 11 above)
Comment (pedromsferraz-mila) [reply src/activelearning/budget/budget_schedule.py:7]: That's a good point, it'd be better to leave this file with just the schedule implementations and move the configs to the separate `budget/config.py` file. Let me know what you think! 1661a3da39826692f85f7522635cdc02700ede96

File: src/activelearning/utils/config_loader.py (hunk 12 above)
Comment (pedromsferraz-mila) [reply src/activelearning/utils/config_loader.py:7]: I agree the naming is a bit confusing here, especially in an ML codebase.

In Pydantic, a “model” means a schema/validated data class: any class inheriting from `pydantic.BaseModel` (https://pydantic.dev/docs/validation/latest/concepts/models/).

`ModelT` is a generic type variable, with `T` following the usual [generic naming convention](https://stackoverflow.com/questions/374014/why-do-generics-often-use-t).

As for `ModelT = TypeVar("ModelT", bound=BaseModel)`, this means `ModelT` can be any subtype of `BaseModel`, so any of our config classes such as `OracleConfig` or `DatasetConfig` fit.

The point is to preserve the concrete config type through helpers like `load_and_parse`:

```
def load_and_parse(
    path: str | Path | list[str | Path],
    model: type[ModelT],
    overrides: list[str] | None = None,
) -> ModelT:
    return parse_config(load_config(path=path, overrides=overrides), model=model)
```

Here, `model: type[ModelT]` means “a class object for some Pydantic model type”, and `-> ModelT` means the function returns an instance of that same type. So if you pass `OracleConfig`, the return type is inferred as `OracleConfig`; if you pass `DatasetConfig`, it is inferred as `DatasetConfig`.

File: src/activelearning/utils/config_loader.py (hunk 13 above)
Comment (pedromsferraz-mila) [reply src/activelearning/utils/config_loader.py:43]: This is used only in tests (to allow more easily loading and parsing configs during testing), and it can also be used as an util for loading configs in scripts or Jupyter notebooks. However, it isn't used in the main active learning loop.

File: src/activelearning/runtime.py (hunk 14 above)
Comment (pedromsferraz-mila) [reply src/activelearning/runtime.py]: Good point, fixed! 3900d32cf4350ee71eaf3b3602f37045d9d51c93

File: src/activelearning/runtime.py (hunk 15 above)
Comment (pedromsferraz-mila) [reply src/activelearning/runtime.py]: Good catch, that was inconsistent. Fixed: 8f952ee3b22c4d36b3ddf726cc951160092df1f0

File: src/activelearning/main.py (hunk 16 above)
Comment (pedromsferraz-mila) [reply src/activelearning/main.py]: Good point! This would work if the env is up-to-date and activated, but it's safer to use `uv run`. Fixed: 5a27c84e48fa7b0bcbf685a0db69cd74ad891c3c

File: src/activelearning/budget/budget_schedule.py (hunk 11 above)
Comment (alexhernandezgarcia) [reply src/activelearning/budget/budget_schedule.py:7]: The changes seem good to me, thanks!

File: src/activelearning/utils/config_loader.py (hunk 12 above)
Comment (alexhernandezgarcia) [reply src/activelearning/utils/config_loader.py:7]: Ok, got it, thanks for the explanation!

**Review by alexhernandezgarcia — APPROVED** [review]
I believe there are no open issues or comments any more, so I'm re-approving this to enable merging.
